### DISCUSSION TITLE
Suppress stack trace for expected exceptions during Service Calls

### DISCUSSION THREAD

**Discussion by jbouwh** [home-assistant/architecture#992] [General] [answered]
## Background

When during the execution of a service call an exception is to be raised due to a user error, e.g.:
- Invalid options supplied
- Invalid entity features
- Invalid config

## Problem

When an exception is raised during a service call and the printing of stack traces is not warranted, this can cause that the logs are being spammed.

## Proposal

1. Introduce `ServiceValidationError` as a new exception type, a sub class of `HomeAssistantError`.

2. We should raise `ServiceValidationError` in service handlers when the user is responsible for the service failing to proceed, either due to invalid service parameters or due to invalid configuration in Home Assistant.

3. The stack trace should be suppressed and an only print at debug level, in case `ServiceValidationError` raised.

4. `HomeAssistantError` and sub classes should allow setting a translation key and translation domain, to allow translations.

## Example

In case the reloading the yaml config fails, we want to omit the stack trace if we can print a clear error message.

PR https://github.com/home-assistant/core/pull/102592 implements the proposal.
PR https://github.com/home-assistant/developers.home-assistant/pull/1961 documents it
PR https://github.com/home-assistant/core/pull/102410  shows demonstrates how the new exception can be used when reloading yaml config was failed.

Update:
- Renamed `ServiceError` to `ServiceValidationError`.
- Added translation key
- Added info about when we should raise `ServiceValidationError`

**Comment by MartinHjelmare** [comment]
The name `ServiceError` doesn't explain what the reason for the error is. Currently the convention is to raise `ValueError` for bad user input and `HomeAssistantError` for api errors in service handlers. If `ServiceError` should replace `ValueError` I think we should have a name that says something about what the reason for the error in the service is.

**Reply by jbouwh** [reply]
@elupus came with `ValidationError`.

**Reply by MartinHjelmare** [reply]
Maybe `ServiceValidationError`?

**Reply by emontnemery** [reply]
> the convention is to raise ValueError for bad user input and HomeAssistantError for api errors in service handlers.

I did not know that's a convention, but it seems reasonable. That convention should be documented somewhere in <https://developers.home-assistant.io/docs/core>

**Reply by jbouwh** [reply]
Updated the proposal

**Reply by emontnemery** [reply]
If we add a `ServiceValidationError`, should all existing `ValueError` raised by services be changes to `ServiceValidationError`? Would a new class `ServiceValidationError` inherit `valueError` or `HomeAssistantError`?

**Reply by elupus** [reply]
My proposal for validation error is not a new global standard exception. It is the specific case that yaml validation fails.

**Reply by jbouwh** [reply]
IMO it should also be usable to raise if a user calls a service with invalid options, where we now use `HomeAssistantError`

**Reply by MartinHjelmare** [reply]
> I did not know that's a convention, but it seems reasonable. That convention should be documented somewhere in https://developers.home-assistant.io/docs/core



It's currently documented here:
https://developers.home-assistant.io/docs/integration_quality_scale_index#silver-

**Reply by MartinHjelmare** [reply]
I think all errors that we raise and expect to happen during normal operation should inherit `HomeAssistantError`.

Sometimes there are cases that could only happen if there is a developer error and then I think it's ok to raise `RuntimeError` or `ValueError` for example.

There are exceptions to the rule where we use a library that has exceptions, like inside voluptuous validators we raise `vol.Invalid`.

**Reply by jbouwh** [reply]
May be we should catch `ValueError` and filter the stack trace when it is raised during a service call?
`vol.Invalid` is already catched and but error is logged.
Suggestion is to catch `ValueError` at `homeassistant/components/websocket_api/commands.py` and suppress the stack trace unless in debug mode. But `ValueError` is a generic error.

**Reply by MartinHjelmare** [reply]
No, I think we should replace raising `ValueError` with `ServiceValidationError`. `ValueError` isn't specific enough and could happen for other reasons. We want a specific exception that we raise which has a specific meaning.

**Comment by emontnemery** [comment]
> > I did not know that's a convention, but it seems reasonable. That convention should be documented somewhere in https://developers.home-assistant.io/docs/core
> 
> It's currently documented here: https://developers.home-assistant.io/docs/integration_quality_scale_index#silver-

I see, I think the quality scale should not be the only place this convention is documented. We should move it somewhere else, and link to there from the quality scale.

**Reply by jbouwh** [reply]
https://github.com/home-assistant/developers.home-assistant/pull/1961 documents the new convention.

**Comment by Jc2k** [comment]
If these are "normal" errors that are somewhat expected to happen, and they are shown in Ui, should we add a translation_key to this new exception?

**Reply by jbouwh** [reply]
Draft PR https://github.com/home-assistant/core/pull/102592 that implements the new exception and adds a translation key. The translation key can be used for `HomeAssistantError` and all subclasses.

**Reply by jbouwh** [reply]
Updated the proposal adding the implementation of a translation key.

**Comment by jbouwh** [comment, marked as answer]
The [proposal](https://github.com/home-assistant/architecture/discussions/992#discussion-5760607) has been updated to make more clear when to raise the new `ServiceValidationError` exception as well as to address the comments by @MartinHjelmare
